tools/flash-shelly.py

```diff
@@ -601,18 +582,51 @@ def device_scan(hosts, action, do_all, dry_run, silent_run, mode, exclude, versi
       deviceinfo.update_to_homekit(homekit_release_info)
     elif flashmode == 'stock':
       deviceinfo.update_to_stock(stock_release_info)
-
     if deviceinfo.fw_type == "homekit" and float(f"{parse_version(deviceinfo.info['version'])[0]}.{parse_version(deviceinfo.info['version'])[1]}") < 2.1:
       logger.info(f"{WHITE}Host: {NC}{deviceinfo.host}")
       logger.info(f"Version {deviceinfo.info['version']} is to old for this script,")
       logger.info(f"please update via the device webUI.\n")
-      continue
-    parse_info(deviceinfo, action, dry_run, silent_run, flashmode, exclude, version, hap_setup_code, requires_upgrade)
+    else:
+      parse_info(deviceinfo, action, do_selection, dry_run, silent_run, flashmode, exclude, hap_setup_code, requires_upgrade)
+
+def device_scan(hosts, action, do_selection, dry_run, silent_run, mode, exclude, version, variant, hap_setup_code):
+  logger.debug(f"\n{WHITE}device_scan{NC}")
+  if hosts:
+    for host in hosts:
+      deviceinfo = Device(host)
+      deviceinfo.get_device_info()
+      if deviceinfo.fw_type is not None:
+        device = {'host': deviceinfo.host, 'wifi_ip': deviceinfo.wifi_ip, 'fw_type': deviceinfo.fw_type, 'device_url': deviceinfo.device_url, 'info': deviceinfo.info}
+        d_info = json.dumps(device, indent = 4)
+        logger.trace(f"Device Info: {d_info}")
+        probe_device(device, action, do_selection, dry_run, silent_run, mode, exclude, version, variant, hap_setup_code)
+  else:
+    logger.info(f"{WHITE}Scanning for Shelly devices...{NC}")
+    zc = zeroconf.Zeroconf()
+    listener = MyListener()
+    browser = zeroconf.ServiceBrowser(zc, '_http._tcp.local.', listener)
+    zc.wait(100)
```

Comment:
no need to wait here

Reply:
the wait i found is required, it is a timeout for browser to return a result, else it can stall and lockup.

```
wait(timeout: float) → None
Calling thread waits for a given number of milliseconds or until notified.
```

Reply:
by this i mean, if a device is online and has an IP, but for some reason it fails to respond to requests, like in my current situation, it just sits there waiting , and waiting, with a 100ms timeout, it silently just ignores the device.

Reply:
missing devices is possible but calling or not calling Zeroconf.wait() has nothing to do with it.
looking at the implementation, it just waits on an internal condition variable.
all the scanning is done in background threads